source/microbit/microbitdisplay.cpp

```diff
@@ -300,14 +301,24 @@ static void microbit_display_update(void) {
              * If an exception is raised here then we must hand it to the VM. */
             mp_obj_t obj;
             nlr_buf_t nlr;
+            gc_lock();
             if (nlr_push(&nlr) == 0) {
                 obj = mp_iternext_allow_raise(async_iterator);
                 nlr_pop();
+                gc_unlock();
             } else {
+                gc_unlock();
                 if (!mp_obj_is_subclass_fast(MP_OBJ_FROM_PTR(((mp_obj_base_t*)nlr.ret_val)->type),
                     MP_OBJ_FROM_PTR(&mp_type_StopIteration))) {
                     // an exception other than StopIteration, so set it for the VM to raise later
-                    MP_STATE_VM(mp_pending_exception) = MP_OBJ_FROM_PTR(nlr.ret_val);
+                    //If memory error, add appropriate message.
+                    mp_obj_exception_t *ex = (mp_obj_exception_t *)nlr.ret_val;
+                    if (mp_obj_get_type(nlr.ret_val) == &mp_type_MemoryError) {
+                        ex->args = (mp_obj_tuple_t *)mp_obj_new_tuple(1, NULL);
```

Comment:
This (and the line below) is allocating memory within an interrupt handler, something we are trying to avoid!  So I'd say just printf the message (actually, use mp_printf(&mp_plat_print, ...)) and let the MemoryError exception through as-is.